### PR TITLE
Add no_std support with optional alloc

### PR DESCRIPTION
**Pull request by GrantM11235**
When the `alloc` feature is disabled, very large words will cause a panic instead of an allocation on the heap, and the `syllables::join` method is disabled. Everything else works exactly the same.

The doc-tests can't use `syllables::join` anymore, so they are a bit noisier. [`itertools::assert_equal`](https://docs.rs/itertools/latest/itertools/fn.assert_equal.html) would help with this, if you are comfortable with adding that as a dev-dependency.

Future possibilities:
- A function that returns an error instead of panicking
- A way to configure the maximum word size with const generics

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -241,14 +263,22 @@ impl Bytes {
...
             Self::Array([0; 40].into_iter(), len)
         } else {
-            Self::Vec(vec![0; len].into_iter())
+            #[cfg(feature = "alloc")]
+            {
+                Self::Vec(vec![0; len].into_iter())
+            }
+            #[cfg(not(feature = "alloc"))]
+            {
+                panic!("Word too large")
```
Comment (laurmaedje) [inline src/lib.rs:272]:
```suggestion
                panic!("word too long");
```

File: src/lib.rs [hunk 2]
```diff
@@ -30,11 +33,16 @@ hypher = { version = "0.1", default-features = false, features = ["english"] }
 ```
 */
 
+#![cfg_attr(not(test), no_std)]
```
Comment (laurmaedje) [inline src/lib.rs:36]: Why is this necessary? Isn't the only thing we need for testing alloc and you're pulling that in below?

File: src/lib.rs [hunk 3]
```diff
@@ -30,11 +33,16 @@ hypher = { version = "0.1", default-features = false, features = ["english"] }
...
 #![forbid(unsafe_code)]
 #![deny(missing_docs)]
 
-use std::fmt::{self, Debug, Formatter};
-use std::iter::FusedIterator;
+#[cfg(any(feature = "alloc", test))]
+#[cfg_attr(not(test), macro_use)]
```
Comment (laurmaedje) [inline src/lib.rs:41]: Also, which macro is this using?

File: src/lib.rs [hunk 4]
```diff
@@ -65,13 +78,20 @@ pub fn hyphenate(word: &str, lang: Lang) -> Syllables<'_> {
...
+///
+/// Panics if the word is more than 38 bytes long and the `alloc` feature is
+/// disabled.
+///
 /// # Example
 /// By setting the left bound to three, we forbid the possible break between
 /// `ex` and `ten`.
 /// ```
 /// # use hypher::{hyphenate_bounded, Lang};
-/// let syllables = hyphenate_bounded("extensive", Lang::English, 3, 1);
```
Comment (laurmaedje) [inline src/lib.rs:73]: This is fine as is, no need to replicate the hack here.

**Review by laurmaedje — CHANGES_REQUESTED** [review]
Thanks for taking a stab at this! If you don't mind me asking, do you actually have a use case for hyphenating on embedded or is this just general no-std-i-fication?

File: src/lib.rs [hunk 5]
```diff
@@ -45,6 +53,11 @@ include!(concat!(env!("OUT_DIR"), "/langs.rs"));
 ///
 /// This uses the default [bounds](Lang::bounds) for the language.
 ///
+/// # Panics
```
[the remark this answers is not in the document]
Comment (GrantM11235) [reply src/lib.rs:56]: I put the "Panics" section before the "Examples" section because that is how it is done in the standard library, but I can still move it if you prefer.

File: src/lib.rs (hunk 3 above)
Comment (GrantM11235) [reply src/lib.rs]: The `vec!` macro. It was only used once, so I just replaced it with `alloc::vec!`

File: src/lib.rs (hunk 2 above)
Comment (GrantM11235) [reply src/lib.rs]: I thought it was necessary, but I was wrong

File: src/lib.rs [hunk 6]
```diff
@@ -14,8 +14,11 @@ _hypher_ separates words into syllables.
 ```rust
 use hypher::{hyphenate, Lang};
 
-let syllables = hyphenate("extensive", Lang::English);
-assert_eq!(syllables.join("-"), "ex-ten-sive");
+let mut syllables = hyphenate("extensive", Lang::English);
+assert_eq!(syllables.next(), Some("ex"));
+assert_eq!(syllables.next(), Some("ten"));
+assert_eq!(syllables.next(), Some("sive"));
+assert_eq!(syllables.next(), None);
```
[the remark this answers is not in the document]
Comment (GrantM11235) [reply src/lib.rs]: I added an additional doctest example for `join` that is only enabled if the `alloc` feature is enabled, based on [this post](https://users.rust-lang.org/t/doctests-that-require-a-non-default-feature-is-it-possible/29529/4).

**Comment by GrantM11235** [comment]
> Thanks for taking a stab at this! If you don't mind me asking, do you actually have a use case for hyphenating on embedded or is this just general no-std-i-fication?

I don't personally have any no_std projects that need hyphenation (not yet anyway), I just thought it was interesting.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (laurmaedje) [reply src/lib.rs:68]: Okay, then it's fine.

**Review by laurmaedje — APPROVED** [review]

**Comment by laurmaedje** [comment]
Thanks again!
